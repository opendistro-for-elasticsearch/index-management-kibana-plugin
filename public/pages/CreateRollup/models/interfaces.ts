/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { ManagedCatIndex } from "../../../../server/models/interfaces";
import { FieldItem } from "../../../../models/interfaces";

export interface MappingsItem {
  properties?: {};
  type?: string;
}

export interface DimensionItem {
  sequence: number;
  field: FieldItem;
  aggregationMethod: string;
  interval?: number;
}

export interface MetricItem {
  source_field: FieldItem;
  all: boolean;
  min: boolean;
  max: boolean;
  sum: boolean;
  avg: boolean;
  value_count: boolean;
}

//From backend
export interface RollupDimensionItem {
  date_histogram?: {
    fixed_interval?: string;
    calendar_interval?: string;
    source_field: string;
    target_field: string;
    timezone: string;
  };
  terms?: {
    source_field: string;
    target_field: string;
  };
  histogram?: {
    source_field: string;
    target_field: string;
    interval: number;
  };
}

export interface RollupMetricItem {
  source_field: string;
  metrics: [
    {
      min?: Object;
      max?: Object;
      sum?: Object;
      avg?: Object;
      value_count?: Object;
    }
  ];
}
