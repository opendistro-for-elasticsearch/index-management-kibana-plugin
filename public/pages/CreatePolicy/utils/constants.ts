/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const DEFAULT_POLICY = JSON.stringify(
  {
    policy: {
      description: "A simple default policy that changes the replica count between hot and cold states.",
      default_state: "hot",
      states: [
        {
          name: "hot",
          actions: [{ replica_count: { number_of_replicas: 5 } }],
          transitions: [{ state_name: "cold", conditions: { min_index_age: "30d" } }],
        },
        {
          name: "cold",
          actions: [{ replica_count: { number_of_replicas: 2 } }],
          transitions: [],
        },
      ],
    },
  },
  null,
  4
);
