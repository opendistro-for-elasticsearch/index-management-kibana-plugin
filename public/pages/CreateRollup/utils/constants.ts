/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

//Added last_updated_time and start_time to avoid error, but they should be system generated.
//Added a delay as 0, but this should be optional, so null should be fine.
export const EMPTY_ROLLUP = JSON.stringify({
  rollup: {
    continuous: false,
    description: "",
    dimensions: [
      {
        date_histogram: {
          source_field: "",
          fixed_interval: "2ms",
          timezone: "Africa/Abidjan",
        },
      },
    ],
    enabled: false,
    metrics: [],
    page_size: 1000,
    roles: [],
    schedule: {
      interval: {
        start_time: 234802,
        period: 1,
        unit: "MINUTES",
      },
    },
    source_index: "",
    target_index: "",
  },
});

export const FixedTimeunitOptions = [
  { value: "ms", text: "Millisecond(s)" },
  { value: "s", text: "Second(s)" },
  { value: "m", text: "Minute(s)" },
  { value: "h", text: "Hour(s)" },
  { value: "d", text: "Day(s)" },
];

export const DelayTimeunitOptions = [
  { value: "SECONDS", text: "Second(s)" },
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const CalendarTimeunitOptions = [
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const AddFieldsColumns = [
  {
    field: "label",
    name: "Field name",
    sortable: true,
  },
  {
    field: "type",
    name: "Field type",
    sortable: true,
    render: (type: string | undefined) => (type == null || type == undefined ? "-" : type),
  },
];
