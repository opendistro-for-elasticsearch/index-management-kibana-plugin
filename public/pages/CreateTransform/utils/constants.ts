/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const EMPTY_TRANSFORM = JSON.stringify({
  transform: {
    description: "",
    enabled: true,
    page_size: 1000,
    roles: [],
    schedule: {
      interval: {
        start_time: 234802,
        period: 1,
        unit: "MINUTES",
      },
    },
    source_index: "",
    target_index: "",
  },
});

export const FixedTimeunitOptions = [
  { value: "ms", text: "Millisecond(s)" },
  { value: "s", text: "Second(s)" },
  { value: "m", text: "Minute(s)" },
  { value: "h", text: "Hour(s)" },
  { value: "d", text: "Day(s)" },
];

export const DelayTimeunitOptions = [
  { value: "SECONDS", text: "Second(s)" },
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const CalendarTimeunitOptions = [
  { value: "m", text: "Minute" },
  { value: "h", text: "Hour" },
  { value: "d", text: "Day" },
  { value: "w", text: "Week" },
  { value: "M", text: "Month" },
  { value: "q", text: "Quarter" },
  { value: "y", text: "Year" },
];

export const ScheduleIntervalTimeunitOptions = [
  { value: "MINUTES", text: "Minute(s)" },
  { value: "HOURS", text: "Hour(s)" },
  { value: "DAYS", text: "Day(s)" },
];

export const AddFieldsColumns = [
  {
    field: "label",
    name: "Field name",
    sortable: true,
  },
  {
    field: "type",
    name: "Field type",
    sortable: true,
    render: (type: string | undefined) => (type == null || type == undefined ? "-" : type),
  },
];

export const ExecutionFrequencyDefinitionOptions = [
  { value: "fixed", text: "Define by fixed interval" },
  { value: "cron", text: "Define by cron expression" },
];

export const DefaultSampleDataSize = 50;
