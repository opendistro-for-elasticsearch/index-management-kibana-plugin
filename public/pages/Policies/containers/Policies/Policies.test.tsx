/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, wait } from "@testing-library/react";
// @ts-ignore
import userEvent from "@testing-library/user-event";
import { Redirect, Route, RouteComponentProps, Switch } from "react-router";
import { MemoryRouter as Router } from "react-router-dom";
import { toastNotifications } from "ui/notify";
import chrome from "ui/chrome";
import { browserServicesMock } from "../../../../../test/mocks";
import Policies from "./Policies";
import { TEXT } from "../../components/PolicyEmptyPrompt/PolicyEmptyPrompt";
import { ModalProvider, ModalRoot } from "../../../../components/Modal";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { BREADCRUMBS, ROUTES } from "../../../../utils/constants";
import { BrowserServices } from "../../../../models/interfaces";

// TODO: Move common renderWith or with___ helpers into top level tests directory
function renderPoliciesWithRouter() {
  return {
    ...render(
      <Router>
        <ServicesContext.Provider value={browserServicesMock}>
          <ServicesConsumer>
            {(services: BrowserServices | null) =>
              services && (
                <ModalProvider>
                  <ModalRoot services={services} />
                  <Switch>
                    <Route
                      path={ROUTES.POLICIES}
                      render={(props: RouteComponentProps) => (
                        <div style={{ padding: "25px 25px" }}>
                          <Policies {...props} policyService={services.policyService} />
                        </div>
                      )}
                    />
                    <Route path={ROUTES.CREATE_POLICY} render={props => <div>Testing create policy</div>} />
                    <Route path={ROUTES.EDIT_POLICY} render={props => <div>Testing edit policy: {props.location.search}</div>} />
                    <Redirect from="/" to={ROUTES.POLICIES} />
                  </Switch>
                </ModalProvider>
              )
            }
          </ServicesConsumer>
        </ServicesContext.Provider>
      </Router>
    ),
  };
}

// TODO: generate random policy (w/ seed for snapshots)
const testPolicy = {
  id: "some_policy_id",
  primaryTerm: 1,
  seqNo: 5,
  policy: {
    policy: {
      description: "some policy description",
      states: [
        {
          name: "some_state",
          actions: [{ delete: {} }],
          transitions: [],
        },
      ],
    },
  },
};

describe("<Policies /> spec", () => {
  it("renders the component", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: [], totalPolicies: 0 } });
    const { container } = renderPoliciesWithRouter();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows LOADING on mount", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: [], totalPolicies: 0 } });
    const { getByText } = renderPoliciesWithRouter();

    getByText(TEXT.LOADING);
  });

  it("sets breadcrumbs when mounting", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: [], totalPolicies: 0 } });
    renderPoliciesWithRouter();

    expect(chrome.breadcrumbs.set).toHaveBeenCalledTimes(1);
    expect(chrome.breadcrumbs.set).toHaveBeenCalledWith([BREADCRUMBS.INDEX_MANAGEMENT, BREADCRUMBS.POLICIES]);
  });

  it("loads policies", async () => {
    const policies = [testPolicy];
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies, totalPolicies: 1 } });
    const { getByText } = renderPoliciesWithRouter();
    await wait();

    await wait(() => getByText(testPolicy.id));
  });

  it("adds error toaster when get policies has error", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    renderPoliciesWithRouter();

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("some error");
  });

  it("adds error toaster when get policies throws error", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockRejectedValue(new Error("rejected error"));
    renderPoliciesWithRouter();

    await wait();

    expect(toastNotifications.addDanger).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addDanger).toHaveBeenCalledWith("rejected error");
  });

  it("can delete a policy", async () => {
    const policies = [testPolicy];
    browserServicesMock.policyService.getPolicies = jest
      .fn()
      .mockResolvedValueOnce({ ok: true, response: { policies, totalPolicies: 1 } })
      .mockResolvedValueOnce({ ok: true, response: { policies: [], totalPolicies: 0 } });
    browserServicesMock.policyService.deletePolicy = jest.fn().mockResolvedValue({ ok: true, response: true });
    const { queryByText, getByText, getByTestId } = renderPoliciesWithRouter();

    await wait(() => getByText(testPolicy.id));

    expect(getByTestId("DeleteButton")).toBeDisabled();

    userEvent.click(getByTestId(`checkboxSelectRow-${testPolicy.id}`));

    expect(getByTestId("DeleteButton")).toBeEnabled();

    userEvent.click(getByTestId("DeleteButton"));

    userEvent.click(getByTestId("confirmationModalActionButton"));

    await wait();

    expect(browserServicesMock.policyService.deletePolicy).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addSuccess).toHaveBeenCalledTimes(1);
    expect(toastNotifications.addSuccess).toHaveBeenCalledWith(`Deleted the policy: ${testPolicy.id}`);

    await wait(() => expect(queryByText(testPolicy.id)).toBeNull());
  });

  it("can route to edit policy", async () => {
    const policies = [testPolicy];
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies, totalPolicies: 1 } });
    const { getByText, getByTestId } = renderPoliciesWithRouter();

    await wait(() => getByText(testPolicy.id));

    expect(getByTestId("EditButton")).toBeDisabled();

    userEvent.click(getByTestId(`checkboxSelectRow-${testPolicy.id}`));

    expect(getByTestId("EditButton")).toBeEnabled();

    userEvent.click(getByTestId("EditButton"));

    await wait(() => getByText(`Testing edit policy: ?id=${testPolicy.id}`));
  });

  it("can route to create policy", async () => {
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies: [], totalPolicies: 1 } });
    const { getByText, getByTestId } = renderPoliciesWithRouter();

    await wait();

    userEvent.click(getByTestId("Create policyButton"));

    await wait(() => getByText("Testing create policy"));
  });

  it("can open and close a policy in modal", async () => {
    const policies = [testPolicy];
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies, totalPolicies: 1 } });
    const { getByText, queryByText, getByTestId } = renderPoliciesWithRouter();

    await wait(() => getByText(testPolicy.id));

    userEvent.click(getByText(testPolicy.id));

    // asserts that the policy description showed up in modal as the
    // whole JSON is broken up between span elements
    await wait(() => getByText(`"${testPolicy.policy.policy.description}"`));

    userEvent.click(getByTestId("policyModalCloseButton"));

    expect(queryByText(`"${testPolicy.policy.policy.description}"`)).toBeNull();
  });

  it("can go to edit a policy from modal", async () => {
    const policies = [testPolicy];
    browserServicesMock.policyService.getPolicies = jest.fn().mockResolvedValue({ ok: true, response: { policies, totalPolicies: 1 } });
    const { getByText, getByTestId } = renderPoliciesWithRouter();

    await wait(() => getByText(testPolicy.id));

    userEvent.click(getByText(testPolicy.id));
    userEvent.click(getByTestId("policyModalEditButton"));

    await wait(() => getByText(`Testing edit policy: ?id=${testPolicy.id}`));
  });

  it("sorts/paginates the table", async () => {
    const policies = new Array(40).fill(null).map((_, index) => ({
      ...testPolicy,
      id: `some_policy_id_${index}`,
      // last_updated_time is effectively adding 1 to every policy so we can assert on the sort
      policy: { policy: { ...testPolicy.policy.policy, last_updated_time: Date.now() + index } },
    }));
    browserServicesMock.policyService.getPolicies = jest
      .fn()
      .mockResolvedValueOnce({ ok: true, response: { policies: policies.slice(0, 20), totalPolicies: 40 } })
      .mockResolvedValueOnce({ ok: true, response: { policies: policies.slice(20), totalPolicies: 40 } })
      .mockResolvedValueOnce({
        ok: true,
        response: {
          policies: policies
            .slice()
            .sort((a, b) => a.policy.policy.last_updated_time - b.policy.policy.last_updated_time)
            .slice(0, 20),
          totalPolicies: 40,
        },
      });

    const { getByText, getByTestId, getAllByTestId, queryByText } = renderPoliciesWithRouter();

    // should load policies 0-19 on first load
    await wait(() => getByText("some_policy_id_0"));
    expect(queryByText("some_policy_id_39")).toBeNull();

    userEvent.click(getAllByTestId("pagination-button-next")[0]);

    // should load policies 20-39 after clicking next
    await wait(() => getByText("some_policy_id_39"));
    expect(queryByText("some_policy_id_0")).toBeNull();

    // @ts-ignore
    userEvent.click(getByTestId("tableHeaderCell_policy.policy.last_updated_time_2").firstChild);

    // should load policies 0-19  after clicking sort (defaults to asc) on last_updated_time
    await wait(() => getByText("some_policy_id_0"));
    expect(queryByText("some_policy_id_39")).toBeNull();
  });
});
