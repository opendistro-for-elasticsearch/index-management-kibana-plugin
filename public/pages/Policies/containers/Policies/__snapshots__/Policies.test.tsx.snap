// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexPolicies /> spec renders the component 1`] = `
<div
  style="padding: 25px 25px;"
>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding-left: 0px; padding-right: 0px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px 10px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--large"
        >
          Index policies
        </h3>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <button
                  class="euiButton euiButton--primary"
                  data-test-subj="DeleteButton"
                  disabled=""
                  type="button"
                >
                  <span
                    class="euiButton__content"
                  >
                    <span
                      class="euiButton__text"
                    >
                      Delete
                    </span>
                  </span>
                </button>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <button
                  class="euiButton euiButton--primary"
                  data-test-subj="EditButton"
                  disabled=""
                  type="button"
                >
                  <span
                    class="euiButton__content"
                  >
                    <span
                      class="euiButton__text"
                    >
                      Edit
                    </span>
                  </span>
                </button>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <button
                  class="euiButton euiButton--primary"
                  data-test-subj="Create policyButton"
                  type="button"
                >
                  <span
                    class="euiButton__content"
                  >
                    <span
                      class="euiButton__text"
                    >
                      Create policy
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: initial;"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        style="padding: 0px 5px;"
      >
        <div
          class="euiFlexItem"
        >
          <div
            class="euiFormControlLayout euiFormControlLayout--fullWidth"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldSearch euiFieldSearch--fullWidth"
                placeholder="Search"
                type="search"
                value=""
              />
              <div
                class="euiFormControlLayoutIcons"
              >
                <span
                  class="euiFormControlLayoutCustomIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                    focusable="false"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <defs>
                      <path
                        d="M11.271 11.978l3.872 3.873a.502.502 0 0 0 .708 0 .502.502 0 0 0 0-.708l-3.565-3.564c2.38-2.747 2.267-6.923-.342-9.532-2.73-2.73-7.17-2.73-9.898 0-2.728 2.729-2.728 7.17 0 9.9a6.955 6.955 0 0 0 4.949 2.05.5.5 0 0 0 0-1 5.96 5.96 0 0 1-4.242-1.757 6.01 6.01 0 0 1 0-8.486c2.337-2.34 6.143-2.34 8.484 0a6.01 6.01 0 0 1 0 8.486.5.5 0 0 0 .034.738z"
                        id="search-a"
                      />
                    </defs>
                    <use
                      xlink:href="#search-a"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
          style="max-width: 250px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiSuperDatePicker__flexWrapper euiSuperDatePicker__flexWrapper--isAutoRefreshOnly"
          >
            <div
              class="euiFlexItem"
            >
              <div
                class="euiFormControlLayout euiFormControlLayout--group euiSuperDatePicker"
              >
                <div
                  class="euiPopover euiPopover--anchorDownLeft"
                  id="QuickSelectPopover"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      aria-label="Date quick select"
                      class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--iconRight euiFormControlLayout__prepend"
                      data-test-subj="superDatePickerToggleQuickMenuButton"
                      type="button"
                    >
                      <span
                        class="euiButtonEmpty__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiButtonEmpty__icon"
                          focusable="false"
                          height="16"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                          xmlns:xlink="http://www.w3.org/1999/xlink"
                        >
                          <defs>
                            <path
                              d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                              id="arrow_down-a"
                            />
                          </defs>
                          <use
                            fill-rule="nonzero"
                            xlink:href="#arrow_down-a"
                          />
                        </svg>
                        <span
                          class="euiButtonEmpty__text euiQuickSelectPopover__buttonText"
                        >
                          <svg
                            class="euiIcon euiIcon--medium"
                            focusable="false"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7.5 13A5.506 5.506 0 0 1 2 7.5C2 4.467 4.467 2 7.5 2S13 4.467 13 7.5 10.533 13 7.5 13m0-12a6.5 6.5 0 1 0 0 13 6.5 6.5 0 0 0 0-13m3 6H8V3.5a.5.5 0 0 0-1 0v4a.5.5 0 0 0 .5.5h3a.5.5 0 0 0 0-1"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <div
                    class="euiDatePickerRange euiDatePickerRange--readOnly euiDatePickerRange--inGroup euiFormControlLayout__child--noStyle"
                  >
                    <span
                      class="euiSuperDatePicker__prettyFormat"
                    >
                      Off
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <hr
        class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
      />
      <div
        class="euiBasicTable"
      >
        <div>
          <div
            class="euiTableHeaderMobile"
          >
            <div
              class="euiTableSortMobile"
            >
              <div
                class="euiPopover euiPopover--anchorDownRight"
                id="sortPopover"
              >
                <div
                  class="euiPopover__anchor"
                >
                  <button
                    class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall euiButtonEmpty--iconRight euiButtonEmpty--flushRight"
                    type="button"
                  >
                    <span
                      class="euiButtonEmpty__content"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiButtonEmpty__icon"
                        focusable="false"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                        xmlns:xlink="http://www.w3.org/1999/xlink"
                      >
                        <defs>
                          <path
                            d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                            id="arrow_down-a"
                          />
                        </defs>
                        <use
                          fill-rule="nonzero"
                          xlink:href="#arrow_down-a"
                        />
                      </svg>
                      <span
                        class="euiButtonEmpty__text"
                      >
                        Sorting
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <table
            class="euiTable euiTable--responsive"
          >
            <caption
              aria-live="polite"
              aria-relevant="text"
              class="euiScreenReaderOnly"
              role="status"
            >
              Below is a table of 0 items.
            </caption>
            <thead>
              <tr>
                <th
                  class="euiTableHeaderCellCheckbox"
                  scope="col"
                  width="24px"
                >
                  <div
                    class="euiTableCellContent"
                  >
                    <div
                      class="euiCheckbox euiCheckbox--inList euiCheckbox--noLabel"
                    >
                      <input
                        aria-label="Select all rows"
                        class="euiCheckbox__input"
                        data-test-subj="checkboxSelectAll"
                        disabled=""
                        id="_selection_column-checkbox"
                        type="checkbox"
                      />
                      <div
                        class="euiCheckbox__square"
                      />
                    </div>
                  </div>
                </th>
                <th
                  aria-live="true"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_id_0"
                  role="columnheader"
                  scope="col"
                  width="150px"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Policy
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  aria-live="true"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_policy.policy.description_1"
                  role="columnheader"
                  scope="col"
                  width="150px"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Description
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
                <th
                  aria-live="true"
                  aria-sort="none"
                  class="euiTableHeaderCell"
                  data-test-subj="tableHeaderCell_policy.policy.last_updated_time_2"
                  role="columnheader"
                  scope="col"
                  width="150px"
                >
                  <button
                    class="euiTableHeaderButton"
                    data-test-subj="tableHeaderSortButton"
                    type="button"
                  >
                    <span
                      class="euiTableCellContent"
                    >
                      <span
                        class="euiTableCellContent__text"
                      >
                        Last updated time
                      </span>
                      <span
                        class="euiScreenReaderOnly"
                      >
                        Click to sort in ascending order
                      </span>
                    </span>
                  </button>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="euiTableRow"
              >
                <td
                  class="euiTableRowCell euiTableRowCell--isMobileFullWidth"
                  colspan="4"
                >
                  <div
                    class="euiTableCellContent euiTableCellContent--alignCenter"
                  >
                    <span
                      class="euiTableCellContent__text"
                    >
                      <div
                        class="euiEmptyPrompt"
                        style="max-width: 45em;"
                      >
                        <span
                          class="euiTextColor euiTextColor--subdued"
                        >
                          <div
                            class="euiText euiText--medium"
                          >
                            <div
                              class="euiText euiText--medium"
                            >
                              <p>
                                Loading policies...
                              </p>
                            </div>
                          </div>
                        </span>
                      </div>
                    </span>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div>
          <div
            class="euiSpacer euiSpacer--m"
          />
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div
                class="euiPopover euiPopover--anchorUpRight euiPopover--withTitle"
                id="customizablePagination"
              >
                <div
                  class="euiPopover__anchor"
                >
                  <button
                    class="euiButtonEmpty euiButtonEmpty--text euiButtonEmpty--xSmall euiButtonEmpty--iconRight"
                    type="button"
                  >
                    <span
                      class="euiButtonEmpty__content"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiButtonEmpty__icon"
                        focusable="false"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                        xmlns:xlink="http://www.w3.org/1999/xlink"
                      >
                        <defs>
                          <path
                            d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                            id="arrow_down-a"
                          />
                        </defs>
                        <use
                          fill-rule="nonzero"
                          xlink:href="#arrow_down-a"
                        />
                      </svg>
                      <span
                        class="euiButtonEmpty__text"
                      >
                        Rows per page
                        : 
                        20
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <span />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
